File: ios/MidTrim/MidTrim/Domain/UseCases/RestoreEntitlementUseCase.swift
import Foundation

struct RestoreEntitlementUseCase {
    private let storeService: StoreKitServiceProtocol
    private let cache: EntitlementCacheProtocol
    private let productID: String

    init(storeService: StoreKitServiceProtocol, cache: EntitlementCacheProtocol, productID: String = "com.midtrim.fullunlock") {
        self.storeService = storeService
        self.cache = cache
        self.productID = productID
    }

    func execute() async -> RestoreResult {
        let result = await storeService.restorePurchases(productID: productID)
        switch result {
        case .found:
            await cache.setPurchased(true)
            await cache.setLastVerified(Date())
        case .notFound:
            await cache.setPurchased(false)
        case .failed:
            break
        }
        return result
    }
}

File: android/app/src/main/java/com/izzatismail/midtrim/domain/usecase/RestoreEntitlementUseCase.kt
package com.izzatismail.midtrim.domain.usecase

import com.izzatismail.midtrim.domain.repository.EntitlementCacheWriter
import com.izzatismail.midtrim.domain.repository.PlayBillingService
import com.izzatismail.midtrim.domain.repository.RestoreResult
import java.util.Date

class RestoreEntitlementUseCase(
    private val billingService: PlayBillingService,
    private val cache: EntitlementCacheWriter
) {
    suspend fun execute(): Boolean {
        val result = billingService.restorePurchases()
        return when (result) {
            is RestoreResult.Found -> {
                cache.isPurchased = true
                cache.lastVerifiedAt = Date().time
                true
            }
            is RestoreResult.NotFound -> {
                cache.isPurchased = false
                false
            }
            is RestoreResult.Failed -> false
        }
    }
}